Add "Messaging API" spec (#19)
waku-org/specs: standards/application/messaging-api.md

##### `staticStoreNode`
A list of `Multiaddr` addresses to remote peers that MUST be used for retrieving past messages or performing infrequent queries using the [STORE](https://github.com/vacp2p/rfc-index/blob/8ee2a6d6b232838d83374c35e2413f84436ecf64/waku/standards/core/13/store.md) protocol.
If not provided, the implementation SHOULD utilize either `preferredServiceNodes` or nodes discovered via the network.

##### `preferredServiceNodes`
A list of `Multiaddr` addresses to remote peers that SHOULD be used for obtaining various network resources.

These resources MAY include:
- infrequent [STORE](../standards/core/store.md) queries;
- a [LIGHTPUSH](../standards/core/lightpush.md) endpoint for broadcasting data;
- a [FILTER](https://github.com/vacp2p/rfc-index/blob/7b443c1aab627894e3f22f5adfbb93f4c4eac4f6/waku/standards/core/12/filter.md) endpoint for receiving circulating data.

If preferred nodes are not provided or are unreachable (i.e., a connection cannot be established due to errors or timeouts),
the implementation SHOULD use nodes discovered through the network.

##### `bootstrapNodes`
This property MUST be provided.

> A list of `Multiaddr` addresses to remote peers that MUST be used for any applicable method of discovery that MAY include:

Reviewer: multiaddrs, enrs or enrtrees

- [DNS Discovery](https://eips.ethereum.org/EIPS/eip-1459);
- [DISCV5](https://github.com/vacp2p/rfc-index/blob/8ee2a6d6b232838d83374c35e2413f84436ecf64/waku/standards/core/33/discv5.md);
- [PEER-EXCHANGE](https://github.com/vacp2p/rfc-index/blob/8ee2a6d6b232838d83374c35e2413f84436ecf64/waku/standards/core/34/peer-exchange.md).

##### `storeConfirmation`
This optional property defaults to `true`.
- If set to `true`, the `stored` field on `MessageRecord` MUST be populated, and implementations MUST employ store‑based reliability as specified in [P2P‑RELIABILITY](./p2p‑reliability.md).
- If set to `false`, the `stored` field on `MessageRecord` MUST NOT be populated.
For more information about the [STORE](../standards/core/store.md) query and the `MessageRecord`, refer to the `Message Storage API` section.

##### `receiveConfirmation`
This optional property defaults to `true`.
- If set to `true`, the node MUST initiate message‑reception detection as specified in [P2P‑RELIABILITY](./p2p‑reliability.md) and populate the `received` field on `MessageRecord`.  
- If set to `false`, the `received` field MUST NOT be populated, unless the `Subscribe API` is invoked.

> 

Reviewer: I don't understand the difference. is `receiveConfirmation` something more general? ie, can either use store or filter?
the name is misleading because "receive" implies that the target recipient received the message which an e2e reliability matter, out of scope for now.

I think I would say that we can have a `networkConfirmation` setting.
And remove `storeConfirmation`. and leave it to the implementation on whether a filter subscription or a store query is used to this confirmation.

##### `confirmContentTopics`
An optional property that SHOULD be provided in conjunction with either `storeConfirmation` or `receiveConfirmation`.
It includes an array of `contentTopics` that the node MUST start monitoring in the background upon initialization,

> as defined in the [TOPICS](https://github.com/vacp2p/rfc-index/blob/8ee2a6d6b232838d83374c35e2413f84436ecf64/waku/informational/23/topics.md#content-topic-format) specification.

Reviewer: I'd remove that. I assume this because you want to be able to setup some filter subscription for content topics the user will send to?
We are making the API more complex here. It seems that filter confirmations are unpractical?

No methods or `REST` endpoints are provided for initial configuration.
Instead, the initial configuration MUST be supplied at node creation time according to the guidelines of the chosen implementation.

### Send

The `Send API` is responsible for broadcasting messages across the network using the configured protocol.
The node SHOULD select the appropriate protocol based on its configuration and the protocols that are mounted:
- If the initial configuration specifies mode: `edge`, then [LIGHTPUSH](../standards/core/lightpush.md) MUST be used to send messages.
- If the initial configuration specifies mode: `relay`, then [RELAY](https://github.com/vacp2p/rfc-index/blob/0277fd0c4dbd907dfb2f0c28b6cde94a335e1fae/waku/standards/core/11/relay.md) MUST be used to send messages.

The `Send API` is also responsible for ensuring message delivery and MUST attempt retries:
- Whenever message loss is detected, in accordance with [P2P‑RELIABILITY](./p2p‑reliability.md).  
- Implementations MAY choose their own backoff timing, maximum number of retry attempts and prioritization of requests.

#### Programmatic API

```typescript
type RequestID = string;

type Message = {
  meta?: UInt8Array;

>   timestamp?: number;

Reviewer: Remove this. The implementation should be the one setting the timestamp.

The method MUST throw an error if the `pubsubTopic` provided by the `Encoder` is not supported by the underlying node.
The method MUST return the associated `RequestID` for the scheduled send attempt.
If the send operation fails, it MAY be retried from the `Message Storage API` using the `RequestID`.
Additionally, send operation can be monitored by subscribing to `message:sent` or `message:error` events described in `Event Source API` below.

##### `cancelRequest`
Cancels a scheduled send attempt associated with a given `RequestID`.
A call to this method MUST be ignored if the request has already been fulfilled (i.e., if it has succeeded, failed, or been canceled previously) or if the provided `RequestID` is invalid.

##### `getPendingRequests`
Returns an array of unique `RequestID` values corresponding to send operations that are still pending.
A request is considered pending if it has not yet succeeded,
and it may be actively retrying or scheduled for sending but has not yet been served.
No duplicate `RequestID` values MUST appear in the returned array.

> 

Reviewer: I'd avoid those if possible. I assume `pending request` is mainly useful for REST API so fine to keep, but maybe we can specify that it's optional to implement in a language API.

- The `RequestID` described in the `Send API` section.
- The `SubscriptionID` described in the `Subscribe API` section.

These identifiers SHOULD be passed to the appropriate methods described below to retrieve or manage messages.

Depending on the presence of the `storeConfirmation` or `receiveConfirmation` options in the `Initial configuration`, the `Message Storage API` will trigger either the `Subscribe API` or `Acknowledgements` routine.

Once a message is added to the `Message Storage API`, or if any details about it change,
the appropriate events from the `message:*` event family (as described in the `Event Source API` section) MUST be invoked after the message is added or updated in the underlying storage.

#### Acknowledgements

Message Acknowledgement mechanism is a background operation that MUST:
- Perform store‑based reliability checks as specified in [P2P‑RELIABILITY](./p2p‑reliability.md).  
- Halt when the node’s health status transitions to `unhealthy` (as defined by the `Health API`).  
- Resume when the node’s health status transitions to either `minimally healthy` or `healthy`.  

> - Set the `stored` or `received` field on `MessageRecord` once the message is acknowledged according to store‑based reliability.

Reviewer: I would not use the term `acknowledgements` as it implies e2e reliability.

I would not list all of those and leave it to implementation detail. just refer to p2p reliability.

- The `storeConfirmation` or `receiveConfirmation` property is `true` and `confirmContentTopics` is provided.  
- The `Subscribe API` is invoked for one or more content topics.  
- The `Send API` has successfully sent one or more messages.

For the recurring [STORE](../standards/core/store.md) queries it is RECOMMENDED to prioritize nodes in this order:
1. `staticStoreNode` specified in the initial configuration.
2. `preferredServiceNodes`, if provided.
3. Nodes discovered via network discovery.

#### Programmatic API

```typescript
type MessageRecord = {
  sending?: boolean;
  sent?: boolean;
  stored: boolean;
  received: boolean;
  requestId?: RequestID;
  message: IWakuMessage;
  error?: unknown;
};

interface MessageStorage {
  getByRequestId(id: RequestID): MessageRecord | undefined;
  getByHash(hash: string, skip?: number, take?: number): MessageRecord | undefined;
  getByContentTopic(contentTopic: string, skip?: number, take?: number): MessageRecord[];
  getBySubscriptionId(id: SubscriptionID, skip?: number, take?: number): MessageRecord[];
}
```

#### Properties of `MessageRecord`

##### `sending`

> Defaults to `false`.

Reviewer: The concept of `default` does not make sense here.
it's either `true` because there's still attempt to send
or `false` because it stops sending due to giving up or being somewhat successful

> Indicates whether a message, originating from the underlying node,

Reviewer: ```suggestion
Indicates whether an outbound message,
```

is currently in the process of being transmitted via the `Send API`.
This property is set to `true` while the message is actively being sent or during retry attempts.
Once the send operation completes (either successfully or if retry attempts are halted),

> the property MUST transition back to `false`.

Reviewer: ```suggestion
the property MUST transition to `false`.
```

> This property is not applicable to messages that are received by the node.

Reviewer: ```suggestion
This property is not applicable to inbound messages.
```

##### `sent`

> Defaults to `false`.

Reviewer: ```suggestion
```

> This field is applicable only to messages sent by the underlying node.

Reviewer: ```suggestion
This field is applicable only to outbound messages.
```

> Indicates whether a message, sent by the underlying node via the `Send API`,

Reviewer: ```suggestion
Indicates whether an outbound message, sent via the `Send API`,
```

was successfully transmitted.
Once a message is sent successfully, the `sent` property becomes `true` and remains unchanged.
If the transmission fails, the property transitions to `false` and the `error` field is populated with implementation-specific details.

> This update occurs only after the `sending` property has transitioned back to `false`.

Reviewer: From what heuristics?

I think some clarification is needed.

I'd say:

Indicates whether a message was sent over-the-wire with a best effort heuristic depending on the outbound protocols used. In the case of Light Push, an OK response from the remote service node will set this property to `true`.
For relay, a seemingly healthy pool of peers will sent this to `true` as soon as the message has been written on the wire.

##### `stored`
Defaults to `false`.
Indicates whether a message has been confirmed to be present in the underlying [STORE](../standards/core/store.md) protocol.
This property is populated based on a recurring [STORE](../standards/core/store.md) query described in `Acknowledgements` section or via the `History API`.
Once a message is confirmed as stored, the `stored` property MUST be updated to `true` as a final state.
If an error occurs during reception, the `error` field MUST be populated with implementation-specific details.

##### `received`
Defaults to `false`.
Indicates whether a message was successfully received from the network via `Subscribe API`.
Once a message is received, the `received` property MUST be updated to `true` as a final state.

> If an error occurs during reception, the `error` field MUST be populated with implementation-specific details.

Reviewer: as discussed earlier, merged those to a  `confirmed` status.

If both `skip` `and` take are omitted, the method MUST return all matching records.
If no records match the specified `contentTopic`, an empty array MUST be returned.

##### `getBySubscriptionId`
Retrieves an array of `MessageRecord` objects corresponding to messages received under the specified `SubscriptionID` (as defined in the `Subscribe API` section).
The method accepts two optional parameters for pagination:
- `skip`: A non-negative integer specifying the number of initial matching records to bypass. If `skip` exceeds the total number of matching records, the method MUST return an empty array.
- `take`: A non-negative integer specifying the maximum number of records to return. If `take` is 0, the method MUST return an empty array. If `take` is greater than the number of available records, all remaining matching records are returned.

If both `skip` and `take` are omitted, the method MUST return all matching records.
If no records match the specified `SubscriptionID`, an empty array MUST be returned.

> 

Reviewer: Too many APIs, especially with `getPendingRequests` already there.

I would say pick ONE way to check on message status at first, and we can then extend as we learn from dogfooding.

```

Response:
```http
HTTP/1.1 200 OK
Content-Type: application/json

[
  {
    "sent": false,
    "stored": false,
    "received": true,
    "message": {
      "payload": "SW5pdGlhbCBtZXNzYWdl",
      "contentTopic": "/messaging/2/api/utf8"
    }
  }
]
```

Error for `404`:
```http
HTTP/1.1 404 Not Found
Content-Type: application/json

{
  "error": "No messages found for subscriptionId '1692854c-cbde-4afe-901f-1c86ecbd9ea1'"
}
```

### Health Indicator

The `Health Indicator API` SHOULD be used to monitor the operational health of a node operating under the `Messaging API`.
The criteria for determining a node’s health state MAY vary depending on the protocols in use and the node’s configuration.
Detailed criteria for each health state are provided in the subsequent sections.

For the purposes of this specification, three health states are defined:
- `unhealthy`: Indicates that the node has lost connectivity for message reception, sending, or both, and as a result, it cannot reliably process or transmit messages.
- `minimally healthy`: Indicates that the node meets the minimum operational requirements, although performance or reliability may be impacted.

> - `healthy`: Indicates that the node is operating optimally, with full support for message processing and transmission.

Reviewer: delete that and just referred to the verbose definition in eponymous sections below

}
```

##### `CustomEvent`
A container object that holds the payload of an event along with implementation-specific metadata.
The `detail` property carries the event payload, while other metadata may vary based on the Waku implementation.

##### `EventHandler`
A callback function that is invoked whenever an event with a supported `eventName` is dispatched.
Event handlers are triggered in the order in which they were added (i.e., in the order of subscription).
The function receives a `CustomEvent` containing the event payload.

##### `addEventListener`
Registers an event listener for a specified `eventName` using the provided `EventHandler`.
If the given `eventName` is not supported, the call MUST be silently ignored.
Additionally, if an exception is thrown within the function passed to `addEventListener`,
that error MUST be ignored and not affect subsequent event handling.

##### `removeEventListener`
Removes a previously registered event listener for the specified `eventName` and `EventHandler`.
If the `eventName` is not supported or the `EventHandler` was not previously registered,
the call MUST be silently ignored.

#### Events

> ##### `health:change`

Reviewer 2: IMO, this should belong to `waku-api`

Refer to `Health Indicator API` section for REST API endpoint.

##### `message:change`
This event MUST be dispatched whenever any of the key properties of a `MessageRecord` change.
Specifically `stored`, `received`,`sent` or `sending`.
The event payload (accessible via `event.detail`) is the updated `MessageRecord` reflecting the changes.

```typescript
node.events.addEventListener("message:change", (event: CustomEvent<MessageRecord>) => {
  // event.detail holds the updated MessageRecord
  console.log(event.detail);
});
```

For the `REST API`, long polling over the `Message Store API` SHOULD be used.

##### `message:sent`
This event MUST be dispatched immediately after a message that originated from the underlying node transitions from being in the `sending` state
(i.e., when `sending` becomes `false`) to a state where it is confirmed as successfully transmitted (i.e., `sent` becomes `true`).
The payload is the `MessageRecord` reflecting this updated status.

```typescript
node.events.addEventListener("message:sent", (event: CustomEvent<MessageRecord>) => {
  // event.detail holds the updated MessageRecord
  console.log(event.detail);
});
```

> For the `REST API`, long polling over the `Message Store API` SHOULD be used.

Reviewer 2: Shall we leave all the `REST` definitions for 2026H1?

---
title: MESSAGING-API
name: Messaging API definition
category: Standards Track
tags: [reliability, application, api, protocol composition]
editor: Oleksandr Kozlov <oleksandr@status.im>
contributors:
- Oleksandr Kozlov <oleksandr@status.im>
- Prem Chaitanya Prathi <prem@status.im>
---

## Abstract

> This document specifies an Application Programming Interface (API) that is RECOMMENDED for consumers of the [WAKU2](https://github.com/vacp2p/rfc-index/blob/7b443c1aab627894e3f22f5adfbb93f4c4eac4f6/waku/standards/core/10/waku2.md) protocol suite as a single entry point to its functionalities.

Author: updated https://github.com/waku-org/specs/pull/19/commits/04cd4bfbcf53669c4007af8fe529b6717de5f0cd

The API generalizes the following core protocols provided by Waku:
- [RELAY](https://github.com/vacp2p/rfc-index/blob/0277fd0c4dbd907dfb2f0c28b6cde94a335e1fae/waku/standards/core/11/relay.md)
- [STORE](../standards/core/store.md)
- [LIGHTPUSH](../standards/core/lightpush.md)

> - [FILTER](https://github.com/vacp2p/rfc-index/blob/7b443c1aab627894e3f22f5adfbb93f4c4eac4f6/waku/standards/core/12/filter.md)

Author: removed this block https://github.com/waku-org/specs/pull/19/commits/04cd4bfbcf53669c4007af8fe529b6717de5f0cd

This API extends the core protocols with additional functionality
required for peer-to-peer messaging applications, including:
- Defines a set of events which MAY be used for subscription to track message propagation, node state, and error handling.
- Defines a set of method calls which MAY be used for invocation of protocol operations.

> - Abstracts new primitives to be accessed in a RESTful manner via [Waku REST API](https://waku-org.github.io/waku-rest-api/).

Author: good point https://github.com/waku-org/specs/pull/19/commits/1e4227ac80a507f09f3f3d2c757277c2206a4e6c

This document defines the specifications and guidelines necessary for implementing the API,
ensuring interoperability and consistency across the Waku protocol family.

## Design Requirements
The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, 
“RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in [RFC2119](https://www.ietf.org/rfc/rfc2119.txt).

### Definitions

This section defines key terms and concepts used throughout this document.
Each term is detailed in dedicated subsections to ensure clarity and consistency in interpretation.

`Multiaddr` - a self-describing format for encoding network address of a remote peer as per [libp2p addressing](https://github.com/libp2p/specs/blob/6d38f88f7b2d16b0e4489298bcd0737a6d704f7e/addressing/README.md) specification.

### Motivation

Real-world application development has exposed challenges in achieving comprehensive usability guarantees for core Waku protocols.

> Although recommendations such as [P2P-RELIABILITY](./p2p-reliability.md) have been introduced to enhance protocol robustness,

Author: improved text here https://github.com/waku-org/specs/pull/19/commits/f1d100fd6c1fa486cd26216dcf10bd4726ae7be2

the fragmented landscape of guidelines and the requirement to re-implement reliability strategies across diverse use cases remain evident.

A significant challenge has been the lack of a unified abstraction for message exchange.
Developers require a straightforward mechanism to send,
receive, and track messages—whether originating from a node or propagated throughout its connected network.

> 

Reviewer 3: Generic Comments:

1. Might be good to have a section listing all API's and referring to subsections for each of them. otherwise for a dev/user it would be hard to navigate the spec to read so much detail. 
2. Also it might be a good idea to provide an API for the user to get localNodeInfo i.e (MultiAddr node is listening on, ENR of the node if applicable). This would be useful in relay node+service node scenarios. So maybe this API is applicable only if node is initialized with relay/serviceNode.

Author: Added table of content https://github.com/waku-org/specs/pull/19/commits/f1d100fd6c1fa486cd26216dcf10bd4726ae7be2

As for (2) - I think it is out of scope of `Messaging API` as this is something that, I think, must be provided by any Waku client. 

We can thought use an opportunity here to specify it. Let me know if you are in favor of it and I will push that change, @chaitanyaprem .

Reviewer 3: > As for (2) - I think it is out of scope of `Messaging API` as this is something that, I think, must be provided by any Waku client.

ok, since we are defining the API i thought we can provide this as well as it might be useful for users to know their nodeinfo. We can add it as it is a small API.

Author: sure, it's only couple of lines
which part of API do you think it fits best, @chaitanyaprem ?

Reviewer 3: I am guessing where we have health indicator.

## API design

### Requirements

> This API is designed for generic use and ease of implementation across `nwaku`, `js-waku`,

Author: https://github.com/waku-org/specs/pull/19/commits/d8796b6a780780411187792e1cf585d1ba3168f6

> This API is designed for generic use and ease of implementation across `nwaku`, `js-waku`,

Author: made it simply `across all implementations and REST`

and traditional `REST` architectures.
From this point forward,

> relevant code blocks will be provided in `JSON` format, `HTTP` format or by using `TypeScript` syntax.

Reviewer: Have you considered moving to a C syntax @weboko ?
I think derivation from C to REST API, TypeScript, etc can then be left to the implementation (ish)

### Initial configuration

The `Messaging API` is an abstraction layer built upon the basic functionality provided by a node.  
The configuration definitions that follow omit implementation-specific details,
focusing exclusively on settings explicitly required by the `Messaging API`.

```typescript
{

>   mode: "edge" | "service";

Author: General question to anyone from me:
- I don't believe we actually need this property AS `Messaging API` can be smart enough to determine and prioritize underlying protocols. 

I am in favor of removing it altogether giving a seamless way for mounting `Messaging API` on any node.

Author: Then I keep `edge` and introduce `relay`. 
As for the `service`, correct me if I am wrong, but I think it is out of the scope of `Messaging API`.

Author: https://github.com/waku-org/specs/pull/19/commits/b6e6e6291fde668ea9d9829c63a762ea02145b62

Reviewer: Are we suggesting that for now, if one wants to run a store node or full blow service node, they would use a binary such a `wakunode2` and not the messaging API?

Sounds fair (for now).

  clusterId: number;
  shards: number[];

>   storeNodes?: string[];

Author: This is interesting design decision I'd like to discuss.

In this proposal I specify following:
`If not provided, nodes discovered through the network SHOULD be used.`

So that, a client implementing `Messaging API` would attempt to recover at least somehow. 
Downside here is that provided service might not be ideal.

Allowing a client to recover can be a way to go as we might have in future stable network, better nodes etc. 
Contrary not providing service is an option as this will make project developers cautious AND willing to do their setup for better overall end experience.

I specified `recovery` approach as this is something, I believe, of a current behavior.

>   storeNodes?: string[];

Reviewer: I agree with what you are saying, I am just saying that in terms of terminology, I believe `staticStoreNode` better convey the functionality you have described. Because you specific static nodes to use, instead of finding nodes in the network (when not specified)

>   storeNodes?: string[];

Author: right :) https://github.com/waku-org/specs/pull/19/commits/b7e0e0b48fb274290e11afef25e6bfcf3f4e1a29

>   serviceNodes?: string[];

Reviewer 4: Would the assumption be that same nodes are used for publishing and receiving (via lightpush and filter)? Since we have storeNodes separate, would it make sense to also have something like `publishNodes`, `listenNodes`?

Reviewer 4: That is my understanding of this field - preferred service nodes, not "the only service nodes to use"

Author: renaming to `preferredServiceNodes` as this is original intention and specified in description already

Author: https://github.com/waku-org/specs/pull/19/commits/e60bd6f6248c0f7d2c5b967ef7910193b3b394c2

  bootstrapNodes: string[];

>   confirmStored?: boolean;

Author: there is an art to naming, and it's hard...

renaming this and `confirmReceived` as I think both must follow the same naming 

the other one is `filterConfirmation` 

https://github.com/waku-org/specs/pull/19/commits/e60bd6f6248c0f7d2c5b967ef7910193b3b394c2

  confirmReceived?: boolean;
  confirmContentTopics?: string[];
}
```

#### Properties

##### `mode`
This property defines behavior of the node and MUST be specified.

If `edge` selected the node MUST use [LIGHTPUSH](../standards/core/lightpush.md) for sending messages,

> and [FILTER](https://github.com/vacp2p/rfc-index/blob/7b443c1aab627894e3f22f5adfbb93f4c4eac4f6/waku/standards/core/12/filter.md) for receiving messages.

Author: @Ivansete-status I am not sure I understand this comment, can you elaborate, please?

> and [FILTER](https://github.com/vacp2p/rfc-index/blob/7b443c1aab627894e3f22f5adfbb93f4c4eac4f6/waku/standards/core/12/filter.md) for receiving messages.

Author: https://github.com/waku-org/specs/pull/19/commits/37d7e7ef52cdd90a4820306114f2d0678e2469ed

> and [FILTER](https://github.com/vacp2p/rfc-index/blob/7b443c1aab627894e3f22f5adfbb93f4c4eac4f6/waku/standards/core/12/filter.md) for receiving messages.

Reviewer 2: > @Ivansete-status I am not sure I understand this comment, can you elaborate, please?

Sure! The `edge` nodes must act as both `lightpush-client` and `filter-client`. In both cases, the `edge` node is not directly mounting `LIGHTPUSH` nor `FILTER` protocols. Instead, the edge node just dial peers that mount those protocols. Therefore, I think is more precise to mention that.

Another option would be something like:

`MUST use [LIGHTPUSH](../standards/core/lightpush.md) as a client and connect to peers that mount LIGHTPUSH`

and

`[FILTER](https://github.com/vacp2p/rfc-index/blob/7b443c1aab627894e3f22f5adfbb93f4c4eac4f6/waku/standards/core/12/filter.md) as a client and connect to peers that mount FILTER`

If `service` selected the node MUST implement [RELAY](https://github.com/vacp2p/rfc-index/blob/0277fd0c4dbd907dfb2f0c28b6cde94a335e1fae/waku/standards/core/11/relay.md),

> [STORE](../standards/core/store.md) as well as host endpoint for [LIGHTPUSH](../standards/core/lightpush.md) and [FILTER](https://github.com/vacp2p/rfc-index/blob/7b443c1aab627894e3f22f5adfbb93f4c4eac4f6/waku/standards/core/12/filter.md).

Reviewer 4: Requiring store does not feel right - Store nodes have different expectations than Lightpush and Filter nodes - those can come and go and still be useful (unless the fluctuation is in order of seconds or minutes), but Store (even with Store sync) should be more stable and available + enabling the protocol puts requirement on hosting the DB. 

I'd personally see store managed separately

Reviewer 4: To add to it, I might be fine if my phone is a `service` node when on Wifi and plugged in the charger, but I would not want it to store GBs of random and "useless" data:)

`edge` mode SHOULD be used if node functions in resource restricted environment,
where as `service` SHOULD be used if node has no hard restrictions.

##### `clusterId`
This property MUST be provided.
It signifies which cluster a node MUST be operating at as per [RELAY-SHARDING](https://github.com/waku-org/specs/blob/186ce335667bdfdb6b2ce69ad7b2a3a3791b1ba6/standards/core/relay-sharding.md).

##### `shards`
This property MUST be provided.

> An array of shard under a specified cluster that node MUST operate at as per [RELAY-SHARDING](https://github.com/waku-org/specs/blob/186ce335667bdfdb6b2ce69ad7b2a3a3791b1ba6/standards/core/relay-sharding.md).

Author: By this description I implied that this is a list of shards that node is interested in, i.e `MUST operate at`
is present description misleading, @chaitanyaprem ?

> An array of shard under a specified cluster that node MUST operate at as per [RELAY-SHARDING](https://github.com/waku-org/specs/blob/186ce335667bdfdb6b2ce69ad7b2a3a3791b1ba6/standards/core/relay-sharding.md).

Reviewer 3: for autosharding logic to work properly there must be a config somewhere to be specified how many shards are there in the cluster. 
wanted to know which config is this one. if it is only the shards the node is interested in, then we may need to specify another config to indicate total shards in the network. 
e.g cluster-id 1 has 8 shards, but node is only interested in shards 0,1 then this config will list [0,1]. 
But somewhere shardCount 8 also has to be specified so that autosharding logic can use that to determine shard based on contentTopic.

> An array of shard under a specified cluster that node MUST operate at as per [RELAY-SHARDING](https://github.com/waku-org/specs/blob/186ce335667bdfdb6b2ce69ad7b2a3a3791b1ba6/standards/core/relay-sharding.md).

Author: actually you are right, I missed it
for some reason I thought all auto shard clusters expected to have 8 shards 
adding an option for it with default value being 8 shards

let me know if you agree with that - https://github.com/waku-org/specs/pull/19/commits/cb446bd873a463abc6edfb212a218798809da2d9

on a side note - it seems `numbers of shards` is only needed for deriving shard from content topic
in that sense, shouldn't it be included into content topic format itself?

> An array of shard under a specified cluster that node MUST operate at as per [RELAY-SHARDING](https://github.com/waku-org/specs/blob/186ce335667bdfdb6b2ce69ad7b2a3a3791b1ba6/standards/core/relay-sharding.md).

Reviewer 3: > on a side note - it seems `numbers of shards` is only needed for deriving shard from content topic
> in that sense, shouldn't it be included into content topic format itself?

well, the idea was to make users agnostic to shards altogether and only use content-topic. but as each cluster can define its own number of shards, it is requiring to be configured locally. Note that the ideal way to get this info would be through some network configuration made maybe inside RLN contract rather than local configuration. As we don't have network configuration specified anywhere as of now, it is required to be configured locally. 

The idea of encoding shard-count in content-topic would  create a dependency of content-topic on shard count. In a cluster ideally the number of shards can change over time based on the network's requirements e.g today cluster 1 has 8 shards based on some napkin math of network requirements and bandwidth usage, but this may get increased in future. If we make content-topic format dependent on shard-count this would create problem to all existing content-topics when shard count changes in the network and would require migration etc.

> An array of shard under a specified cluster that node MUST operate at as per [RELAY-SHARDING](https://github.com/waku-org/specs/blob/186ce335667bdfdb6b2ce69ad7b2a3a3791b1ba6/standards/core/relay-sharding.md).

Reviewer 3: > let me know if you agree with that - [cb446bd](https://github.com/waku-org/specs/commit/cb446bd873a463abc6edfb212a218798809da2d9)

maybe the order and naming of config could be something like this as clusterId and shardsInCluster are network level config which are not specific to node being initialized whereas `shardsOfInterest` is something specific to the local node.

```
clusterId:
shardsInCluster:
shardsOfInterest:
```

##### `storeNodes`
A list of `Multiaddr` addresses to remote peers that SHOULD be used for retrieving past messages or performing infrequent queries using the [STORE](https://github.com/vacp2p/rfc-index/blob/8ee2a6d6b232838d83374c35e2413f84436ecf64/waku/standards/core/13/store.md) protocol.
If not provided, nodes discovered through the network SHOULD be used.

##### `serviceNodes`
A list of `Multiaddr` addresses to remote peers that SHOULD be used for obtaining various network resources.

These resources MAY include:

> - infrequent [STORE](../standards/core/store.md) queries;

Author: yes, my intent to have following approach for store:
- if `storeNodes` provided they are used else:
- if `preferredServiceNodes` they are used else:
- try to use discovered nodes at this point:
- or fail.

This is a design decision to make here, I describe it here. 
Comment what you think is better - https://github.com/waku-org/specs/pull/19/files#r2023781573

- a [LIGHTPUSH](../standards/core/lightpush.md) endpoint for broadcasting data;
- a [FILTER](https://github.com/vacp2p/rfc-index/blob/7b443c1aab627894e3f22f5adfbb93f4c4eac4f6/waku/standards/core/12/filter.md) endpoint for receiving circulating data.

> If not provided, nodes discovered through the network SHOULD be used.

Author: good point, added this change
https://github.com/waku-org/specs/pull/19/commits/f6516dad45eafcbd1e5b46d690e3396019540fbb

##### `bootstrapNodes`
This property MUST be provided.

> A list of `Multiaddr` addresses to remote peers that SHOULD be used for any applicable method of discovery that MAY include:

Reviewer 3: wondering if we should also allow the option to specify `ENR` in case of bootstrap/store/service nodes?

- [EIP-1459](https://eips.ethereum.org/EIPS/eip-1459);
- [DISCV5](https://github.com/vacp2p/rfc-index/blob/8ee2a6d6b232838d83374c35e2413f84436ecf64/waku/standards/core/33/discv5.md);
- [PEER-EXCHANGE](https://github.com/vacp2p/rfc-index/blob/8ee2a6d6b232838d83374c35e2413f84436ecf64/waku/standards/core/34/peer-exchange.md).

> ##### `confirmStored`

Reviewer 3: maybe call it as `sendConfirmations`?

Author: very difficult, named it `sotreConfirmation` to give some hint about protocol used too
https://github.com/waku-org/specs/pull/19/commits/e60bd6f6248c0f7d2c5b967ef7910193b3b394c2

Reviewer 3: wondering if giving protocol name is a good idea, anyways since this is supposed to be abstracting protocols would it be better to call it `sendConfirmations` ? as this flag is used only to confirm via store messages which are sent out, better to indicate that in the name.

> An optional property that defaults to `false`.

Reviewer 3: maybe this should be set to true by default for edge nodes. 
maybe check the current default behaviour with status team and align with it as i am not aware of this.

Author: asked question, for now assuming it to be default `on` so defining it to be `true`
https://github.com/waku-org/specs/pull/19/commits/cd55755ffc4e7209dc7640c709619a806af1a39e

> If set to `true` and `confirmContentTopics` are provided, a recurring background [STORE](../standards/core/store.md) query (as described in the `Message Storage API` section) MUST be initiated.

Reviewer 3: Do we want to specify interval of these queries as well based on what is used in case of status (i.e 3 secs)?

Author: @chaitanyaprem specified it here 
https://github.com/waku-org/specs/pull/19/commits/face579a70ff8b1cbf0c1aec18908960495812b5

Reviewer 3: > meanwhile isn't 3 seconds is a short window?
> I thought of 10s or 30s and only if messages were `LightPushed` since we'll have active `Filter` subscription

iirc i think initially it was higher(maybe 5 or 10secs initially) but since status UI uses store confirmation to show message-status to the user 3sec was optimal to not have bad ux.

> If set to `true` without `confirmContentTopics`, the background [STORE](../standards/core/store.md) query (as described in the `Messaging Storage API` section) MUST only be initiated after the `Subscribe API` is called.

Reviewer 3: A user may not always subscribe to content-topics that they are publishing to. Hence this query is supposed to be based on `messageHashes` and not `contentTopics`. 
ContentTopics are to be used in-conjunction with filter only where user subscribes to them to receive messages.

 This field can be used along with `confirmReceived`  rather than `confirmStored`

Author: this is a good point, I misses this case completely
here is definition for this behavior https://github.com/waku-org/specs/pull/19/commits/ed996f3aa5ea310df03bdf18336adb2ed9f332cc

If set to `false`, the `stored` property on the `MessageRecord` (defined in the `Message Storage API`) MUST NOT be populated, unless `History API` is triggered.

> ##### `confirmReceived`

Author: https://github.com/waku-org/specs/pull/19/commits/e60bd6f6248c0f7d2c5b967ef7910193b3b394c2

An optional property that defaults to `true`.

> If set to `true`, the node MUST initiate a network listener for messages circulated under `confirmContentTopics` via either [FILTER](https://github.com/vacp2p/rfc-index/blob/7b443c1aab627894e3f22f5adfbb93f4c4eac4f6/waku/standards/core/12/filter.md) or [RELAY](https://github.com/vacp2p/rfc-index/blob/0277fd0c4dbd907dfb2f0c28b6cde94a335e1fae/waku/standards/core/11/relay.md) using the `Subscribe API`.

Author: Initially, I aimed to keep this spec self-contained,
but after revisiting p2p-reliability, I realize it makes more sense to link them properly.

Leaving this comment open until I find a good way to handle the integration.

> If set to `true`, the node MUST initiate a network listener for messages circulated under `confirmContentTopics` via either [FILTER](https://github.com/vacp2p/rfc-index/blob/7b443c1aab627894e3f22f5adfbb93f4c4eac4f6/waku/standards/core/12/filter.md) or [RELAY](https://github.com/vacp2p/rfc-index/blob/0277fd0c4dbd907dfb2f0c28b6cde94a335e1fae/waku/standards/core/11/relay.md) using the `Subscribe API`.

Author: aligned with p2p-reliability spec in https://github.com/waku-org/specs/pull/19/commits/1e4227ac80a507f09f3f3d2c757277c2206a4e6c

If set to `true` but `confirmContentTopics` are not provided, the received property on the `MessageRecord` (as defined in the `Message Storage API`) MUST be populated only after the `Subscribe API` is called.
If set to `false`, the received property MUST NOT be populated, unless `Subscribe API` is called.

##### `confirmContentTopics`
An optional property that SHOULD be provided in conjunction with either `confirmStored` or `confirmReceived`.
It includes an array of `contentTopics` that the node MUST start monitoring in the background upon initialization,
as defined in the [TOPICS](https://github.com/vacp2p/rfc-index/blob/8ee2a6d6b232838d83374c35e2413f84436ecf64/waku/informational/23/topics.md#content-topic-format) specification.

> 

Reviewer 3: I would suggest this field to be used only for receive message confirmations. See reasoning in above comment.

#### Programmatic API / REST API

No methods or `REST` endpoints are provided for initial configuration.
Instead, the initial configuration MUST be supplied at node creation time according to the guidelines of the chosen implementation.

> ### Send

Reviewer 4: SHould it also mention what happens when the node is in `Unhealthy` state? Or is that implicit - sending will fail and it will be retried when the health changes to `minimally healthy` or `healthy`?

Author: it's specified in `Retry` section below, if it is not clear, please, re-open this comment

The `Send API` is responsible for broadcasting messages across the network using the configured protocol.
The node SHOULD select the appropriate protocol based on its configuration and the protocols that are mounted:
- If the initial configuration specifies mode: "edge", then [LIGHTPUSH](../standards/core/lightpush.md) MUST be used to send messages.

> - If the initial configuration specifies mode: "service", then [RELAY](https://github.com/vacp2p/rfc-index/blob/0277fd0c4dbd907dfb2f0c28b6cde94a335e1fae/waku/standards/core/11/relay.md) MUST be used to send messages.

Reviewer: I would not re-state that, it's infer from the mode of operation section

The `Send API` is also responsible for ensuring message delivery and MUST attempt retries as described in the `Retry` section.

#### Retry

If a message fails to satisfy any of the following conditions, it MUST be automatically resent:
- The message is not successfully sent using [LIGHTPUSH](../standards/core/lightpush.md) or [RELAY](https://github.com/vacp2p/rfc-index/blob/0277fd0c4dbd907dfb2f0c28b6cde94a335e1fae/waku/standards/core/11/relay.md).
- The message is not acknowledged by `Message Storage API` described below.
- Retry attempts MUST NOT continue if the node’s health state changes to unhealthy (as defined in the `Health API` described below).
- Retry attempts MUST resume when the node’s health state transitions to either `minimally healthy` or `healthy` defined in the `Health API` below.

Retry attempts MUST occur in the background using a fixed retry count and exponential backoff, as follows:
- A total of 3 retry attempts MUST be made following the initial send attempt.
- The retry intervals MUST be 1 second, then 2 seconds, and finally 4 seconds, respectively.

> - The decision to retry MUST be based on the underlying protocol’s response from a remote peer.

Reviewer: Two comments here:

1. This can be refer back to the p2p reliability spec
> The usual remedial action is to retransmit the message. 

I would not re-state this logic here. But instead, just clarifying that `send` API is expected to implement the strategies configured and stated in p2p reliability state.

2. Actual back-off timing can be left to implementater's choice

Reviewer: If there is one thing you should include or that cna be configurable is for how long do we retry - because this may be application dependent. but for now, I'll leave it to the implementation

Author: also agree, addressed here https://github.com/waku-org/specs/pull/19/commits/1e4227ac80a507f09f3f3d2c757277c2206a4e6c

#### Request prioritization

Send requests MUST be prioritized based on their submission time and retry status.
Freshly scheduled requests MUST be served before those undergoing retry attempts.
Among retry attempts, processing MUST occur in descending order of recency—that is,

> from the most recently scheduled or retried request to the oldest.

Reviewer: I would not include that here. Implementation detail.

Author: done https://github.com/waku-org/specs/pull/19/commits/1e4227ac80a507f09f3f3d2c757277c2206a4e6c

> from the most recently scheduled or retried request to the oldest.

Author: This should be handled by the node implementing `Messaging API` after `Send API` was triggered. 
Does that answer your question, @Ivansete-status ?

#### Programmatic API

```typescript
type RequestID = string;

> type Message = {

Reviewer: See my comment on writing it in C.

Do the same as https://github.com/vacp2p/rfc-index/blob/main/waku/standards/core/36/bindings-api.md#jsonmessage-type

  meta?: UInt8Array;
  timestamp?: number;
  ephemeral?: boolean;
  payload: UInt8Array;
};

interface ISend {

>   public send(encoder: Encoder, message: Message): RequestID;

Reviewer: Not convinced encoder concept works here. Let's see how it looks on C API.

Reviewer 3: Why do we need to have an encoder? Shouldn't encoder be more like a utility from waku/sdk app devs can use to encode their payload and then supply WakuMessage to be sent? This gives flexibility for devs to decide whether to use Waku encoder or some other way to encode their messages.

This looks like better separate of concerns in case apps want to use their own way to encode different message types e.g: in status we have private messages which are encrypted/encoded differently and public messages which are done differently. Ultimately status backend handsover only WakuMessage to the send API which seems more cleaner to me.

>   public cancelRequest(requestId: RequestID): void;

Reviewer: Do we need/use that? sounds too low level

Reviewer 3: Agreed it is too low level, rather a more simpler approach would be for the user to indicate a timeout for the send API. The user can also pass a context(this might be too Golang specific but is a very good concept for an API in general)  which can be cancelled or timeout.

Author: I thought of having it as a backup for consumers to have granular control 
but after all discussions with you I think maybe some `start` and `stop` methods will be enought

Reviewer: let's do simple first. https://www.notion.so/Waku-Mission-and-Methodology-1a58f96fb65c80b9b789c1bbb9e99915?pvs=4#1d78f96fb65c802c928af2576c55bc20

So yes, start/stop sounds fair

>   public getPendingRequests(): RequestID[];

Reviewer 4: Should there also be `isPending(requestId: RequestID)` method, or is iterating over `getPendingRequests()` result enough?

Reviewer: Ok very interesting here.

Due to the nature of Waku's domain (async network request), it seems much more appropriate to have an event based approach.

See https://github.com/vacp2p/rfc-index/blob/main/waku/standards/core/36/bindings-api.md#waku_set_event_callback

having `getRequestState` implies someone is going to loop on that - not a great design.

I'd remove in favor of event listening/callbacks

Reviewer 3: I agree with @fryorcraken here. It would be better to have an event based approach to indicate success/failure/timeout etc.

};
```

##### `Encoder`
The `Encoder` object encapsulates the logic for message handling.
It MUST ensure that messages are encoded into the appropriate format for transmission.

> Additionally, it MUST provide information regarding the `pubsubTopic` and `contentTopic` to which messages SHOULD be sent.

Reviewer: No mention of `pubsubtopic` in the API, we only deal with shards and clusters.

The specific types of encoders, their extended functionalities, and the mechanisms for their instantiation are considered out of scope for this document.

##### `Message`
A `Message` object MUST conform to the [MESSAGE](https://github.com/vacp2p/rfc-index/blob/8ee2a6d6b232838d83374c35e2413f84436ecf64/waku/standards/core/14/message.md) specification.
- The payload property MUST be provided.
- All other properties are optional; if omitted, they SHOULD be supplied by the `Encoder`.

##### `RequestID`
A `RequestID` is a `GUID` string that uniquely identifies a message send operation.

> This `GUID` MAY be used to cancel a scheduled request or to retrieve information about the associated message from the `Message Storage API`.

Reviewer 3: Why not use the term `messageHash` as it is used in Store API and in all other existing specs?

Author: @chaitanyaprem 

since [message hash](https://github.com/vacp2p/rfc-index/blob/main/waku/standards/core/14/message.md#deterministic-message-hashing) takes into account `timestamp` - due to retries it can change over time

hence we need a better way to identify requests TO `Messaging API` FOR sending a message
then we have a way to match `RequestID` with end `message hash`

This was my reasoning here, let me know what you think!

Reviewer 3: ok, i see what you mean. I am wondering this may pose challenge in quering message from store as user won't have messageHash to work with or use SDS (as i think it relies on messageHashes).

But then again if the idea is to abstract away both of these from user, then it is fine i guess. But this is more of a transactionID between user and local waku messaging API and nothing beyond that right.

Reviewer: I think it's fine for the request id to be purely implementation specific, and mainly used to track events.

If the send operation fails, it MAY be retried from the `Message Storage API` using the `RequestID`.
Additionally, send operation can be monitored by subscribing to `message:sent` or `message:error` events described in `Event Source API` below.

##### `cancelRequest`
Cancels a scheduled send attempt associated with a given `RequestID`.
A call to this method MUST be ignored if the request has already been fulfilled (i.e., if it has succeeded, failed, or been canceled previously) or if the provided `RequestID` is invalid.

##### `getPendingRequests`
Returns an array of unique `RequestID` values corresponding to send operations that are still pending.
A request is considered pending if it has not yet succeeded,
and it may be actively retrying or scheduled for sending but has not yet been served.
No duplicate `RequestID` values MUST appear in the returned array.

#### REST API

##### `send`
Request:
```http
POST /send HTTP/1.1
Accept: application/json
Content-Type: application/json

{
  "timestamp": 0,
  "ephemeral": false,
  "meta": "ZXhhbXBsZQ==",
  "pubsubTopic": "string",
  "contentTopic": "string",
  "payload": "ZXhhbXBsZQ=="
}
```

> 

Reviewer: I would not define things twice. Only once in `C` and provide a recommendation on how this should translate to a REST API

Reviewer: @weboko thoughts on this?

IMO

define a C API:

```
char* waku_send(char* message, int shard, int timeoutMs, WakuCallBack onOkCb, WakuCallBack onErrorCb){}
```

and define a way to deduce the REST endpoint:
1. name function because route, `_` becomes `/`
2. data arguments are properties of the JSON payload
3. bytes payload are encoded in base64
4. callbacks are properties on the response JSON/websocket

Which would be:

``
POST /waku/send

{
	message: WakuMessageJSON,
	shard: number,
}
```

etc

Response:
```http
HTTP/1.1 200 OK
Content-Type: application/json

{
  "requestId": "9978c4f1-4465-4537-9d95-0cf016fb1080"
}
```

Example for `404`:
```http
HTTP/1.1 404 Not Found
Content-Type: application/json

{
  "error": "Failed to send message. Target pubsubTopic 'string' not supported."
}
```

##### `cancelRequest`
Request:
```http
POST /send/cancel HTTP/1.1
Accept: application/json
Content-Type: application/json

{
  "requestId": "9978c4f1-4465-4537-9d95-0cf016fb1080"
}
```

Response:
```http
HTTP/1.1 200 OK
Content-Type: application/json

{
  "status": "ok"
}
```

> ##### `getPendingRequests`

Author: I hope it can be resolved with `Message Storage API` which gives a way to fetch `MessageRecord` with associated `RequestID`. 

And I am not sure if it is the best strategy to do.

> ##### `getPendingRequests`

Author: exactly! I am trying to enclose this logic in `Message Storage API` for `get` operations and relevant events in `Event Source API`

Request:
```http
GET /send/requests HTTP/1.1
Accept: application/json
```

Response:
```http
HTTP/1.1 200 OK
Content-Type: application/json

[
  "9978c4f1-4465-4537-9d95-0cf016fb1080"
]
```

### Subscribe

The `Subscribe API` SHOULD be used to initiate a subscription over [FILTER](https://github.com/vacp2p/rfc-index/blob/7b443c1aab627894e3f22f5adfbb93f4c4eac4f6/waku/standards/core/12/filter.md) or [RELAY](https://github.com/vacp2p/rfc-index/blob/0277fd0c4dbd907dfb2f0c28b6cde94a335e1fae/waku/standards/core/11/relay.md) for continuous message retrieval.

When the `Subscribe API` is invoked, the following behaviors MUST occur.

> A subscription is established via [FILTER](https://github.com/vacp2p/rfc-index/blob/7b443c1aab627894e3f22f5adfbb93f4c4eac4f6/waku/standards/core/12/filter.md) or [RELAY](https://github.com/vacp2p/rfc-index/blob/0277fd0c4dbd907dfb2f0c28b6cde94a335e1fae/waku/standards/core/11/relay.md) based on the node’s configuration for ongoing message reception.

Reviewer 3: Is this a good place to mention that subscription should be done to 2 filter service nodes?

Reviewer: I think we should refer to p2p reliability spec instead, and not add such implementation detail here.

Author: simplified this section a bit https://github.com/waku-org/specs/pull/19/commits/1e4227ac80a507f09f3f3d2c757277c2206a4e6c and https://github.com/waku-org/specs/pull/19/commits/4523673bdc9365580eec792059898d504cfccb90

If `confirmStored` is set to `true`,
a recurring [STORE](../standards/core/store.md) query MUST be initiated for the specified `contentTopics` to continuously update the stored property in the `MessageRecord` (as defined in the `Message Storage API`) for messages circulating in the network after the `Subscribe API` call.
For details on the recurring query, refer to the `Message Storage API` section.

Once the `Subscribe API` is called, the received property in the `MessageRecord` (as described in the `Message Storage API`) MUST be populated.

When the subscription is active, the API MUST trigger the `message:*` event group as described in the `Event Source API` section.

#### Programmatic API

```typescript
type SubscriptionID = string;

type SubscriptionListener = (message: MessageRecord) => void | Promise<void>;

interface ISubscribe {

>   public subscribe(decoder: Decoder, fn: SubscriptionListener): SubscriptionID;

Reviewer: I would remove decoder.

  public unsubscribe(id: SubscriptionID): void;
  public getActiveSubscriptions(): SubscriptionID[];
}
```

##### `SubscriptionID`
A unique identifier (GUID string) representing a subscription.
This ID MUST be used to manage active subscriptions,
allowing them to be terminated when they are no longer needed.

##### `SubscriptionListener`
A callback function that is invoked whenever a new `MessageRecord` is received.
It processes the incoming message and MAY return a `Promise` to support asynchronous handling.

> The `MessageRecord` is defined in the `Message Storage API` section.

Reviewer 3: Would be good to add a note that the callback is expected to not do any long processing so that underlying threads don't get blocked and drops messages. 

We have noticed this issue in status once where message's were getting dropped as the app thread was blocked due to some sync logic which was causing reliability issues.

##### `subscribe`
Registers a new subscription to listen for messages processed by a given `Decoder`.

> The subscription is based on the `pubsubTopic` and `contentTopic` provided by the `Decoder`,

Reviewer 4: As I mentioned above - would it make sense to move away from `pubsubTopic` in dev/user facing parts and only refer to `shard`?

> The subscription is based on the `pubsubTopic` and `contentTopic` provided by the `Decoder`,

Reviewer 3: In case of relay contentTopic would be optional, whereas in case of Filter it is mandatory. Would be good to document his in the API so that users are aware based on the nodeType how the params are to be passed.

Reviewer: > In case of relay contentTopic would be optional, whereas in case of Filter it is mandatory. 

Well, ideally one uses autosharding, making it the same interface for both relay and filter :)

which determine the scope of the subscription and the messages to be received.
The subscription is uniquely identified by a `SubscriptionID`,
which MUST be used for managing the subscription (e.g., for later unsubscribing).
The method MUST throw an error if the `pubsubTopic` provided by the `Decoder` is not supported by the underlying node.

> No specific order is guaranteed for callback invocation;

Reviewer 3: Maybe better to add a note that users would need to have a way to deduplicate messages (returned by 2 filter subscriptions or store nodes etc). Or is that implemented underneath this API? If so, better to mention that as well for implementers.

Reviewer: yes I agree here, we may need to define a deduplication logic on payload within a timeframe.

subscription callbacks MUST be executed as messages are received from the protocol in use.

> If `confirmStored` is set to `true`, a recurring [STORE](../standards/core/store.md) query MUST be started.

Reviewer 4: I understood `confirmStored` as targetting sent messages (i.e. to confirm a message I published reached the network). 

What if I don't necessarily need `confirmStore` for sending (e..g I am just publishing an "I am online" heartbeat, so I don't really care if they are missed sometime), but I do care for regular store queries as I'd really like to make sure I get all the messages? 

Basically, are we conflating 2 separate features under `confirmStored`?

Reviewer 3: I agree and think that we should have 2 configs as it is currently implemented in status. We have separate flag to indicate `sendConfirmations` vs `fetchMissingMessages`

##### `unsubscribe`
Removes a previously created subscription identified by its unique `SubscriptionID`.
Once unsubscribed, the associated `SubscriptionListener` MUST NOT receive any further messages.
If the provided `SubscriptionID` does not correspond to an active subscription,

> the call MUST be ignored.

Reviewer 3: wouldn't it be better to return an error indicating the same?

Reviewer: I think ignore is fine, because you may to unsubscribe for a clean shutdown but if the remote node already unsubscribe you because you were disconnected or something, you dont' want to be collecting errors.

If recurring [STORE](../standards/core/store.md) query was initiated by this `SubscriptionID`, it MUST be stopped.

##### `getActiveSubscriptions`
Returns an array of unique `SubscriptionID` values representing all currently active subscriptions.
Each identifier corresponds to a subscription that is in effect,
and no duplicate values MUST be present in the returned array.

> ##### `Decoder`

Reviewer 4: This depends on how smart the Messaging API should be - e.g. in Waku Dispatcher my "onMsgReceived` function signature looks like this: https://github.com/vpavlin/waku-dispatcher/blob/main/src/dispatcher.ts#L135 - I kind of like the flexibility of using Encoder/Decoder gives us and potentially gives all the devs (I could implement my own Decoder which would do some smart validation or processing)

Another thing is you cannot just pass payload and contentTopic - the dev also needs the rest of the fields - like `meta`, `timestamp` or `rln proof`:)

> ##### `Decoder`

Reviewer 2: > I kind of like the flexibility of using Encoder/Decoder gives us and potentially gives all the devs

ah okay I see. I have the impression that the Encoder/Decoder concept fits very well from typescript PoV but I wonder how it would fit from a C API PoV

----
> Another thing is you cannot just pass payload and contentTopic - the dev also needs the rest of the fields - like meta, timestamp or rln proof:)

Yes good point! I agree that `meta` and `timestamp` is needed but maybe `rln proof` not needed

> ##### `Decoder`

Reviewer 3: > Another thing is you cannot just pass payload and contentTopic - the dev also needs the rest of the fields - like `meta`, `timestamp` or `rln proof`:)

Why not pass WakuMessage itself which has all these fields?

> ##### `Decoder`

Reviewer: I would move away from the Decoder in the context of the spec.

But js-waku is welcome to have an abstraction specific to JS/TS over the messaging API is decoder/encoder is a better dev ex.

The `Decoder` object encapsulates the logic for processing received messages.
It MUST ensure that incoming messages are decoded from their received format into a usable structure.

> Additionally, it MUST filter messages based on the `contentTopic` to ensure that only relevant messages are processed.

Reviewer 3: I am confused about Decoder now. Since it is passed in Susbcribe API i assume that it expects messages belonging to the specific contentTopic and only does actual decoding of the message and not filtering of it. 

What if an app users wants to use different decoders for different contentTopics? Then would each Decoder receive messages of all topics that needs to be filtered? or am i missing something here.

Reviewer: `Decoder` was mainly created to cope with the `version=1` concept, which itself, should also go away :)

Author: I wanted to bring concept of `Decoder` as a convenient abstraction but I now think it's more of `implementation details` and nothing to do with the spec

will remove it

The specific types of decoders, their extended functionalities,
and their instantiation mechanisms are considered out of scope for this document.

#### REST API

Reading messages is performed via the `/messages?subscriptionId="<GUID>"` endpoint,
as detailed in the `Message Storage API` section.

##### `subscribe`
Request:
```http
POST /subscribe HTTP/1.1
Accept: application/json
Content-Type: application/json

{
  "pubsubTopic": "/waku/2/rs/0/2",
  "contentTopic": "/example/2/rfc/proto"
}
```

Response:
```http
HTTP/1.1 200 OK
Content-Type: application/json

{
  "subscriptionId": "9978c4f1-4465-4537-9d95-0cf016fb1080"
}
```

Error for `404`:
```http
HTTP/1.1 404 Not Found
Content-Type: application/json

{

>   "error": "Cannot create subscription. clusterId 0 is not supported."

Reviewer 4: There is no `clusterId` in `/subscribe` - this message would utterly confuse me as a developer:) Probably related to my question(s) about replacing `pubsubTopic` with `shard` (and then also adding `clusterid` to `/subscribe` for completeness)

}
```

##### `unsubscribe`
Request:
```http
POST /unsubscribe HTTP/1.1
Accept: application/json
Content-Type: application/json

{
  "subscriptionId": "9978c4f1-4465-4537-9d95-0cf016fb1080"
}
```

Response:
```http

> HTTP/1.1 200 OK

Reviewer 4: Should there also be `404` or if the subscription does not exist, or would it return `200` since it is `ok`, kinda

Content-Type: application/json

{
  "status": "ok"
}
```

##### `getActiveSubscriptions`
Request:
```http
GET /subscriptions HTTP/1.1
Accept: application/json
```

Response:
```http
HTTP/1.1 200 OK
Content-Type: application/json

[
  "9978c4f1-4465-4537-9d95-0cf016fb1080",
  "a1b2c3d4-5678-90ab-cdef-1234567890ab"
]
```

> ### Message Storage

Reviewer 4: I think we need some Message Storage API:-) The naming might need to be tweaked, but as a dev, I want to be able to sift through the messages I have sent and received, to filter them, or verify things - and it would be bad if everyone had to implement this on their own - Messaging API is a great place for it:)  (speaking again based on what I needed in Waku Dispatcher - https://github.com/vpavlin/waku-dispatcher/blob/main/src/dispatcher.ts#L482)

> ### Message Storage

Reviewer 3: How about we have this API only to do time-range based retrieval queries where a node has come back from offline state and wants history.

The hash based queries can be done behind the scenes as part of the Messaging API for reliability checks.

> ### Message Storage

Reviewer: I think we are talking about local storage. Hence, i would rename to `Local Message Storage API`.

I would NOT include this at this point in time, and would instead aim for the scope to be narrower and spec leaner.

**However**, I believe that the opposite is missing. A Waku implementation needs access to a local DB or cache to function properly, and remember message ids. etc.
This should be defined here. See how go-waku interfaces with status-go DB to mark messages as sent etc.

So when you start a node, there is a couple of components you should pass to your node, and we need to define the APIs those component must implement.
Another one is a `time source`, and `DNS source` (or maybe just DNS Servers ips or DoH addresses).

> ### Message Storage

Reviewer 2: @vpavlin
> I think we need some Message Storage API:-)

Yes, I completely agree that a `history-API` is needed but I consider it belongs to the `waku-api` , instead of the `messaging-api`. That `history-API` may also implement internally the local cache storage mentioned on other comments.

The `Message Storage API` SHOULD be used to retrieve information about messages that have been sent, 
received, or stored.

> Message storage is populated by:

Reviewer 3: Not sure why we want to explain Store functionality here or is this storage something else?

- Messages sent by the underlying node via the `Send API`.
- Messages received through an active subscription initiated by the `Subscribe API`.
- Messages fetched using the `History API`.

Messages can be identified by:
- The `message hash` as described in the [MESSAGE](https://github.com/vacp2p/rfc-index/blob/8ee2a6d6b232838d83374c35e2413f84436ecf64/waku/standards/core/14/message.md#deterministic-message-hashing) specification.
- The `contentTopic` associated with the message, as defined in the [TOPICS](https://github.com/vacp2p/rfc-index/blob/main/waku/informational/23/topics.md#content-topic-format) specification.
- The `RequestID` described in the `Send API` section.
- The `SubscriptionID` described in the `Subscribe API` section.

These identifiers SHOULD be passed to the appropriate methods described below to retrieve or manage messages.

> Depending on the presence of the `confirmStored` or `confirmReceived` options in the `Initial configuration`, the `Message Storage API` will trigger either the `Subscribe API` or the periodic [STORE](../standards/core/store.md) query as described in the `Acknowledgements` section below.

Reviewer 3: Now i am totally confused here. Why is Message Storage API triggering a subscribe?

Once a message is added to the `Message Storage API`, or if any details about it change,
the appropriate events from the `message:*` event family (as described in the `Event Source API` section) MUST be invoked after the message is added or updated in the underlying storage.

#### Acknowledgements

> Message acknowledgement is a background operation that performs the following functions:

Reviewer 3: Maybe refer this section above in confirm flags indicating the mechanism is described here.

I got confused at first wondering if this is in addition to periodic store query mechanism based on `confirm* `flags.

> - It subscribes to [FILTER](https://github.com/vacp2p/rfc-index/blob/7b443c1aab627894e3f22f5adfbb93f4c4eac4f6/waku/standards/core/12/filter.md) or [RELAY](https://github.com/vacp2p/rfc-index/blob/0277fd0c4dbd907dfb2f0c28b6cde94a335e1fae/waku/standards/core/11/relay.md) and marks messages as `received` when detected.

Reviewer 3: why is a background operation subscribing to Filter/relay? isn't it the otherway around i.e if a user subscribes to filter and confirmReceived is set then those shards/content-topics are used for periodic message retrieval checks.

- It periodically queries [STORE](../standards/core/store.md) and marks messages as `stored` when available.

This operation MUST be initiated if either of the following conditions is met:
- Either `confirmStored` or `confirmReceived` is set to `true` and `confirmContentTopics` is provided.
- The `Subscribe API` has been initiated for one or more `contentTopics`.

The recurring [STORE](../standards/core/store.md) query MUST prioritize the `storeNodes` specified in the initial configuration before using other available nodes.

> These queries SHOULD occur once every 60 seconds for all `confirmContentTopics` provided as part of the initial configuration.

Reviewer 3: I think for sendConfirmations a shorter interval is used in case of status iirc it is `3seconds.`

If no records match the specified `SubscriptionID`, an empty array MUST be returned.

#### REST API

##### By `RequestID`
Request:
```http
GET /message?requestId=xyz789 HTTP/1.1
Accept: application/json
```

Response:
```http
HTTP/1.1 200 OK
Content-Type: application/json

{
  "sending": true,
  "sent": false,
  "stored": false,
  "received": false,
  "requestId": "xyz789",
  "message": {
    "payload": "U29tZSBvdGhlciBkYXRh",
    "contentTopic": "/example/another-topic"
  }
}
```

Error for `404`:
```http
HTTP/1.1 404 Not Found
Content-Type: application/json

{
  "error": "Message with requestId 'xyz789' not found"
}
```

##### By hash
Request:
```http
GET /message?hash=abc123 HTTP/1.1
Accept: application/json
```

Response:
```http
HTTP/1.1 200 OK
Content-Type: application/json

{
  "sending": false,
  "sent": false,
  "stored": true,
  "received": false,
  "message": {
    "payload": "SGVsbG8gd29ybGQ=",
    "contentTopic": "/example/topic"
  }
}
```

Error for `404`:
```http
HTTP/1.1 404 Not Found
Content-Type: application/json

{
  "error": "Message with hash 'abc123' not found"
}
```

##### By `contentTopic`
Request:
```http

> GET /messages?contentTopic=/messaging/2/api/utf8&skip=0&take=10 HTTP/1.1

Reviewer 4: Should this be URL encoded?

Content-Type: application/json

{
  "error": "No messages found for contentTopic '/messaging/2/api/utf8'"
}
```

##### By `subscriptionId`
Request:
```http
GET /messages?subscriptionId=1692854c-cbde-4afe-901f-1c86ecbd9ea1&skip=0&take=10 HTTP/1.1
Accept: application/json
```
Note: When omitting pagination parameters, the URL would be:
```http
GET /messages?subscriptionId=1692854c-cbde-4afe-901f-1c86ecbd9ea1 HTTP/1.1
```

Response:
```http
HTTP/1.1 200 OK
Content-Type: application/json

[
  {
    "sent": false,
    "stored": false,
    "received": true,
    "message": {
      "payload": "SW5pdGlhbCBtZXNzYWdl",
      "contentTopic": "/messaging/2/api/utf8"
    }
  }
]
```

Error for `404`:
```http
HTTP/1.1 404 Not Found
Content-Type: application/json

{
  "error": "No messages found for subscriptionId '1692854c-cbde-4afe-901f-1c86ecbd9ea1'"
}
```

### Health Indicator

The `Health Indicator API` SHOULD be used to monitor the health of a node operating under the `Messaging API`.
The specific criteria for determining the health state MAY vary depending on the protocols the node utilizes.

We define three health states:
- unhealthy;
- minimally healthy;
- healthy;

> The `Health Indicator API` SHOULD be used to monitor the operational health of a node operating under the `Messaging API`.

Reviewer 3: Here to 773 seems duplicate of 758-764.

Author: indeed, removed

The criteria for determining a node’s health state MAY vary depending on the protocols in use and the node’s configuration.
Detailed criteria for each health state are provided in the subsequent sections.

For the purposes of this specification, three health states are defined:

> - `unhealthy`: Indicates that the node is in a degraded state and may not reliably process or transmit messages.

Reviewer 3: Doesn't this mean that node cannot process/transmit rather than being in degraded state. This indicates loss of connectivity for reception or sending or both.

Author: agree, improved wording here 
https://github.com/waku-org/specs/pull/19/commits/f401e5842d64b18f8e5fe34707a8f9d30a267fb4

- `minimally healthy`: Indicates that the node meets the minimum operational requirements, although performance or reliability may be impacted.
- `healthy`: Indicates that the node is operating optimally, with full support for message processing and transmission.

Furthermore, any change in the node’s health state MUST trigger a `health:change` event via the `Event Source API` immediately after the state transition occurs.
Refer to `Event Source API` for more details.

#### Health states

##### `unhealthy`
Indicates that the node’s connectivity is insufficient for reliable operation.
In this state:
- No connections to service nodes are available, regardless of protocol.
- No peers are present in the [RELAY](https://github.com/vacp2p/rfc-index/blob/0277fd0c4dbd907dfb2f0c28b6cde94a335e1fae/waku/standards/core/11/relay.md) mesh.

> ##### `minimally healthy`

Reviewer 3: which state would the node be in if it has connections to filter and lightpush but no connection to store nodes?
I am guessing. minimally healthy. maybe better to mention that explicitly as well.

Author: good question, I am not sure
`Store` is important but it doesn't actually cause functionality looses for a node, unless it is expected to get confirmations

so I didn't add it in the description here

Reviewer: I would leave it to implementation details for now, until we have deduced strong heuristics across all implementations.

I would just say that while a node may be able to send or receive messages, the redundancy dictated by the relay and p2p reliability protocols are not inplace.

Indicates that the node meets the minimum connectivity requirements necessary to function,
although performance or reliability may be compromised.
In this state:
- The node has at least one connection to a service node implementing [FILTER](https://github.com/vacp2p/rfc-index/blob/7b443c1aab627894e3f22f5adfbb93f4c4eac4f6/waku/standards/core/12/filter.md) and one connection to a service node implementing [LIGHTPUSH](../standards/core/lightpush.md).
- The [RELAY](https://github.com/vacp2p/rfc-index/blob/0277fd0c4dbd907dfb2f0c28b6cde94a335e1fae/waku/standards/core/11/relay.md) mesh includes connections to at least four other peers.

##### `healthy`
Indicates that the node’s connectivity meets or exceeds optimal operational thresholds.
In this state:
- The node has at least two connections to service nodes implementing [FILTER](https://github.com/vacp2p/rfc-index/blob/7b443c1aab627894e3f22f5adfbb93f4c4eac4f6/waku/standards/core/12/filter.md) and at least two connections to service nodes implementing [LIGHTPUSH](../standards/core/lightpush.md).

> - The [RELAY](https://github.com/vacp2p/rfc-index/blob/0277fd0c4dbd907dfb2f0c28b6cde94a335e1fae/waku/standards/core/11/relay.md) mesh includes connections to at least six other peers.

Reviewer: I think too detailed here, and I would instead say that the redundancy stipulated by the p2p reliability and relay protocls are fulfilled.

Consumers MAY subscribe to this event source to monitor message propagation,
network changes, or the health status of the node.
Events are dispatched only after an actual change to the underlying data occurs,
ensuring that notifications reflect genuine updates.

For accessing events via the REST API, long polling SHOULD be used.

#### Programmatic API

```typescript
type EventHandler = (event: CustomEvent<unknown>): void;

interface IEvents {
  public addEventListener(eventName: string, fn: EventHandler): void;
  public removeEventListener(eventName: string, fn: EventHandler): void;
}
```

##### `CustomEvent`
A container object that holds the payload of an event along with implementation-specific metadata.
The `detail` property carries the event payload, while other metadata may vary based on the Waku implementation.

##### `EventHandler`
A callback function that is invoked whenever an event with a supported `eventName` is dispatched.
Event handlers are triggered in the order in which they were added (i.e., in the order of subscription).
The function receives a `CustomEvent` containing the event payload.

##### `addEventListener`

> Registers an event listener for a specified `eventName` using the provided `EventHandler`.

Reviewer 3: Shouldn't eventListeners be automatically registered for many cases? 
i.e all message related events should be notified to the user such as sent/stored etc. 
wondering why a user should explicitly subscribe to such events?

Author: I am not sure it is possible, to me it seems consumer of the API must somehow pass a reference to `EventHandler` they have, and this is something that can't be done automatically 
unless we expose new parameter in `Initial Configuration` for setting a callback

Reviewer 3: ok..got it. i was thinking of receiving because anyways callback would be provided for filter messages. 
But for sending side user may need to register explicitly or pass a callback while invoking send.

If the given `eventName` is not supported, the call MUST be silently ignored.
Additionally, if an exception is thrown within the function passed to `addEventListener`,
that error MUST be ignored and not affect subsequent event handling.

##### `removeEventListener`
Removes a previously registered event listener for the specified `eventName` and `EventHandler`.
If the `eventName` is not supported or the `EventHandler` was not previously registered,
the call MUST be silently ignored.

#### Events

##### `health:change`
This event MUST be emitted whenever the node’s health state changes,
as determined by the `Health Indicator API`.

> The event’s payload (accessible via `event.detail`) is the new `HealthStatus` value (`Unhealthy`, `MinimallyHealthy`, or `Healthy`).

Reviewer 3: I am wondering if we should indicate store not reachable/not working as a separate state. Because in this case the node is connected to appropriate serviceNodes/relay-mesh but just that store confirmations/retrieval is not working. This seems like it is falling in between minimallyHealthy and Healthy. 

I am also not sure if we need a separate status for this but better to think about it.

Author: That's difficult question.

On one side we want to provide granular way for consumers of the API to react to underlying changes,
on the other side if we expose too much consumers might get lost.

Also with Store it is more complicated as we want to hide it as much as possible.
And on top of that, even if we provide all the health status changes - what reaction do we anticipate?

In present format `HealthManager` covers UI changes so that:
- `unhealthy` -> notify user they are offline;
- `minimally healthy` -> raise a warning that some features may not work BUT rely on underlying guarantees described in the spec, e.g when back online to re-send messages, re-query Store, re-subscribe to Filter;
- `healthy` -> proceed with regular execution;

Maybe instead of exposing `StoreMissing` we should describe another mechanism that will ensure that:
- if node was online but skipped Store queries that were scheduled - it MUST recover messages for that time window;

wdyt @chaitanyaprem and maybe @vpavlin ?

Author: specified message recovery after offline here https://github.com/waku-org/specs/pull/19/commits/face579a70ff8b1cbf0c1aec18908960495812b5
